accounts/abi/bind: Go ABI binding generator

karalabe opened the pull request:
This PR is an initial stab at a web3 equivalent Go binding generator. It creates a `abigen` command that takes a contract ABI and input and generates a Go package/file as output containing all **type safe** wrapper code to interact with an instance of the particular contract on the blockchain.

Some of the features it includes:

 * Wrap all contract methods in type safe Go API functions
 * Have separate types (also) for read-only (free) and write-only (paid) calls/transactions
 * Have the possibility to fins tune all call/transaction parameters (e.g. gas, price, acc, nonce)
 * Implement session types that have the above parameters pre-configured (i.e. simpler API)
 * By providing the bytecode, contract deployment methods are also generated
 * By providing the solidity sources, everything is automagically generated in one go

Beside the binding generation, the PR includes:

 * RPC based contract backend to allow running bindings outside of Geth
 * Simulated contract backend to allow testing contracts and/or bindings

**Logs/events will be implemented in a follow up PR to try and keep the scope of this one a bit more limited. It's already quite large and I see no need to push even more features in, especially since some required dependencies in the RPC side is still pending review.**

robotally commented:
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Wed Mar 16 16:00:13 UTC 2016_

obscuren commented on accounts/abi/abi.go:
```diff
@@ -193,8 +173,8 @@ var (
-// unmarshal output in v according to the abi specification
-func (abi ABI) unmarshal(v interface{}, name string, output []byte) error {
+// Unpack output in v according to the abi specification
+func (abi ABI) Unpack(v interface{}, name string, output []byte) error {
```
Why did you go with `Pack` & `Unpack` instead of `Marshal` & `Unmarshal`? (I'm aware `Pack` was already called `Pack` and this was IMO a mistake).

obscuren replied:
Not  saying that refactoring is wrong, but it could perhaps gone outside of this PR?

karalabe replied:
The reason I didn't do in a separate PR was that it's simple enough and it would have just needed an extra round of reviews and rebases to get merged in. If you want though I can easily separate out these changes and open an explicit PR with them.

obscuren commented on accounts/abi/bind/backend.go:
```diff
@@ -0,0 +1,59 @@
+// to the transactor to decide.
+type ContractTransactor interface {
+	// Nonce retrieves the current pending nonce associated with an account.
+	AccountNonce(account common.Address) (uint64, error)
```
If this is the pending nonce it should perhaps be called `PendingAccountNonce` to better reflect it's behaviour?

karalabe replied:
Fixed, will push in next commit.

obscuren commented on accounts/abi/bind/backend.go:
```diff
@@ -0,0 +1,59 @@
+	// of a transaction.
+	GasPrice() (*big.Int, error)
+
+	// GasLimit tries to estimate the gas needed to execute a specific transaction.
```
How? Or is it intentionally left up to the implementation?

karalabe replied:
Well, implementation wise the RPC backend delegates it to `eth_estimateGas`, the simulated backend executes the transaction on a temp environment directly. I don't think it's really necessary to define how to calculate it (although there's arguable a single sensible way to do it). The only important thing is for the method to return the amount of gas it needs (in the current pending state) to execute the transaction.

obscuren replied:
> The only important thing is for the method to return the amount of gas it needs (in the current pending state) to execute the transaction.

My point being exactly. This method is called `GasLimit` not `estimate gas`. The method doc doesn't really reflect the method correctly, or better put, the method name doesn't really reflect well enough what that limit is *supposed* to be.

karalabe replied:
Fair enough, will change to `EstimateGas` or something similar.

karalabe replied:
Renamed and re-doc-ed to:

```go
	// SuggestGasPrice retrieves the currently suggested gas price to allow a timely
	// execution of a transaction.
	SuggestGasPrice() (*big.Int, error)

	// EstimateGasLimit tries to estimate the gas needed to execute a specific
	// transaction based on the current pending state of the backend blockchain.
	// There is no guarantee that this is the true gas limit requirement as other
	// transactions may be added or removed by miners, but it should provide a basis
	// for setting a reasonable default.
	EstimateGasLimit(sender common.Address, contract *common.Address, value *big.Int, data []byte) (*big.Int, error)
```

will push in next commit.

obscuren commented on accounts/abi/bind/backends/remote.go:
```diff
@@ -0,0 +1,202 @@
+	if err := json.Unmarshal(res, &hex); err != nil {
+		return 0, err
+	}
+	return new(big.Int).SetBytes(common.FromHex(hex)).Uint64(), nil
```
You can use `big.Int`'s [`SetString`](https://golang.org/pkg/math/big/#Int.SetString) method. It accepts hexadecimal strings (or use `common.StringToBig`)

karalabe replied:
Actually, neither `SetString` or `common.String2Big` (which uses `SetString` internally) is appropriate, as the result of the RPC call is prefixed with `0x`, which `SetString` cannot parse: http://play.golang.org/p/lqdQPiPCiP .

obscuren replied:
It can http://play.golang.org/p/3bKving9gB

karalabe replied:
Ah, didn't know about 0. Will fix it up,

karalabe replied:
Changed all occurrences to +-

```go
	price, ok := new(big.Int).SetString(hex, 0)
	if !ok {
		return nil, fmt.Errorf("invalid price hex: %s", hex)
	}
	return price, nil
```

will push in next commit.

obscuren commented on accounts/abi/bind/backends/simulated.go:
```diff
@@ -0,0 +1,184 @@
+		data:     data,
+	}
+	// Execute the call and return
+	vmenv := core.NewEnv(statedb, b.blockchain, msg, block.Header())
```
Just a FYI, this function has changed in https://github.com/ethereum/go-ethereum/pull/2141. See: https://github.com/ethereum/go-ethereum/pull/2141/files#diff-0a440a2dbb68b1d118735c02285f90b0R56

karalabe replied:
Ah thanks. I'll rebase once your linked PR is merged in.

obscuren commented on accounts/abi/bind/backends/simulated.go:
```diff
@@ -0,0 +1,184 @@
+
+// Transact implements ContractTransactor.SendTransaction, delegating the raw
+// transaction injection to the remote node.
+func (b *SimulatedBackend) SendTransaction(tx *types.Transaction) error {
```
General observation:

This feature is now implement 3 times. Here, in the `miner` package using the worker and the `core` package. We'd want to refactor this and make this generic. The `worker`, imo, takes the right approach by taking the block and simply applying the transaction on to the pending state rather than - as you do here - create a new block and reapplying the transactions.

I also understand this is non-critical code but I'd just like to mention it.

karalabe replied:
The reason I did it like this is because I reused `core.ChainMaker`, which already implements almost all of the block generation and transaction processing logic. However, since it pregenerates the blocks and doesn't support further modifying them, my choices were to do a bit of extra work (which given that it's a tester, shouldn't matter that much), or on the other hand duplicate almost all the tx processing and block assembly code yet again here too.

obscuren commented on accounts/abi/bind/base.go:
```diff
@@ -0,0 +1,178 @@
+// deployment address with a Go wrapper.
+func DeployContract(opts *TransactOpts, abi abi.ABI, bytecode []byte, backend ContractBackend) (*BoundContract, error) {
+	// Sanity check the authorization options
+	if opts == nil {
```
If this isn't optional, shouldn't it just panic? i.e. remove the `nil` check.

karalabe replied:
Dunno honestly. I'm not too fond of hard panicking when it can be avoided.

obscuren commented on accounts/abi/bind/base.go:
```diff
@@ -0,0 +1,178 @@
+		return nil, errors.New("transaction options missing")
+	}
+	// Otherwise try to deploy the contract
+	c := NewBoundContract(common.Address{}, abi, backend.(ContractCaller), backend.(ContractTransactor))
```
The casts aren't necessary here since the `ContractBackend` embeds the `ContractCaller` and `ContractTransactor` directly.

karalabe replied:
Ah, indeed, good catch!

obscuren commented on accounts/abi/bind/base.go:
```diff
@@ -0,0 +1,178 @@
+// value as the fund transfer to the contract.
+func (c *BoundContract) Transact(opts *TransactOpts, method string, params ...interface{}) (*types.Transaction, error) {
+	// Sanity check the authorization options
+	if opts == nil {
```
Same here

obscuren replied:
Maybe some rational, it's not like a user, in the code, will first try it without options, checks the error and the figured it needs to pass it and so recalls this function. That's silly. IMO it should hard crash telling the user YOU'RE DOING IT WRONG!

karalabe replied:
Fair enough, I'll change it.

obscuren commented on accounts/abi/bind/base.go:
```diff
@@ -0,0 +1,178 @@
+	if value == nil {
+		value = new(big.Int)
+	}
+	nonce := uint64(0)
```
Could you place a comment here saying something like "If no nonce was specified we'll attempt to guess one using ... etc".

obscuren commented on accounts/abi/bind/bind.go:
```diff
@@ -0,0 +1,395 @@
+// to be used as is in client code, but rather as an intermediate struct which
+// enforces compile time type safety and naming convention opposed to having to
+// manually maintain hard coded strings that break on runtime.
+func Bind(abijson string, bytecode string, pkg string, kind string) (string, error) {
```
Could you write `abijson` as:
* ABIJson
* abiJSON
* abiJson

Sorry for nitpicking :-D

karalabe replied:
`abiJSON`, fixed. You're right, not nitpicking :)

obscuren commented on accounts/abi/bind/bind.go:
```diff
@@ -0,0 +1,395 @@
+	kind = strings.ToUpper(kind[:1]) + kind[1:]
+	fmt.Fprintf(code, "%s\n", bindContract(kind, strings.TrimSpace(abijson), strings.TrimSpace(bytecode)))
+
+	methods := make([]string, 0, len(abi.Methods))
```
Maybe this should be a method on the `ABI` struct itself?

karalabe replied:
Generating a binding? Not sure about it. Originally the entire thing was in ABI, that could have worked, but it starts introducing weird dependencies. The ABI package should really only have to do with calling conventions, code generation is a bit pushing it. I think having `bind` as a subpackage should be enough to make it clear that it's *part* of the ABI, but not an essential thing for usual use cases.

obscuren commented on accounts/abi/bind/bind.go:
```diff
@@ -0,0 +1,395 @@
+	code := ""
+
+	// Generate the hard coded ABI and byte code used for Ethereum interaction
+	code += fmt.Sprintf("// Ethereum ABI used to generate the binding from.\nconst %sABI = `%s`\n\n", kind, abijson)
```
Wouldn't it be much more readable if you used package `text/template` here? This would allow you to do something like:

```go
// Ethereum ABI used to generate the
const {{kind}}ABI = {{abi_json}}

{{if bytecode}}
// Ethereum VM bytecode ..
const {{kind}}ABI = {{bytecode}}
{{end}}

... and so forth ...
```

karalabe replied:
Yes, I'll rework the entire code to use templates. Didn't think that it will get so ugly so fast. My bad :)

karalabe commented:
@obscuren Ready for a second round of review :)

obscuren commented on accounts/abi/bind/backends/simulated.go:
```diff
@@ -0,0 +1,184 @@
+		block = b.blockchain.CurrentBlock()
+		statedb, _ = b.blockchain.State()
+	}
+	statedb = statedb.Copy()
```
There's no need to call `Copy()` because `b.blockchain.State()` will already create a copy of the state by calling `state.New(block.Root(), db)`.

karalabe replied:
Good catch. Still needed for the pending state, but that can go in the first branch only.

obscuren commented on accounts/abi/bind/bind.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+// capitalize makes the first character of a string upper case.
+func capitalize(input string) string {
```
capitaliSe please

karalabe replied:
Fixed.

obscuren commented on accounts/abi/bind/bind.go:
```diff
@@ -0,0 +1,162 @@
+
+// capitalize makes the first character of a string upper case.
+func capitalize(input string) string {
+	return strings.ToUpper(input[:1]) + input[1:]
```
FYI: this will panic on empty string.

karalabe replied:
I know, but it should never be called for the empty string :D That usually means we forgot to handle a possible missing param somewhere, which will cause bigger issues later on anyway.

obscuren replied:
Ok so panic == good thing here. Cool.

obscuren replied:
Actually, hang on. What if we integrate this in to `geth` with the version contract and accidentally screw up ourself with some corner case and cause a massive outage. Would this be possible?

I'm all for letting crucial parts of the application hard fail, but things like the version contract shouldn't ever be able to crash on corner case like there IMO.

karalabe replied:
This code is only ever called when you generate a wrapper. It's never used inside the generated code.

karalabe commented:
Pushed all fixes, apart the nil capitaliSe. I still think it's useful for that to crash since it means the generator did something stupid. As it's not reused inside generated code, just for creating the wrapper, it should be fine to crash, cannot happen runtime in live code.

obscuren commented:
Cool @karalabe :+1: LGTM!

bas-vk commented on cmd/abigen/main.go:
```diff
@@ -0,0 +1,80 @@
+	binFlag = flag.String("bin", "", "Path to the Ethereum contract bytecode (generate deploy method)")
+	pkgFlag = flag.String("pkg", "", "Go package name to generate the binding into")
+	typFlag = flag.String("type", "", "Go struct name for the binding (default = package name)")
+	outFlag = flag.String("out", "", "Output path for the generated binding (default = stdout)")
```
Slightly more accurate usage for outFlag would be:
`Output file for the generated binding (default = stdout)`

karalabe replied:
Fixed and pushed. Thanks!

bas-vk commented on accounts/abi/bind/base.go:
```diff
@@ -0,0 +1,174 @@
+// TransactOpts is the collection of authorization data required to create a
+// valid Ethereum transaction.
+type TransactOpts struct {
+	Account common.Address // Ethereum account to send the transaction from
```
Would prefer `Account` to be renamed to `From` which is better self explanatory and more in line with the conventions used in the RPC layer.

karalabe replied:
Not entirely sure about that. In the case of plain transactions it makes sense because you have a sender and a recipient. Here the recipient is implicitly defined as the receiver contract you are calling transacting methods on, so there can only ever be a single account.

bas-vk replied:
I still prefer the use of from ;-)
Guess its a matter of personal preference.

karalabe replied:
I'll sleep on it ;)

On Wed, Mar 23, 2016 at 7:35 PM, bas-vk <notifications@github.com> wrote:

> In accounts/abi/bind/base.go
> <https://github.com/ethereum/go-ethereum/pull/2357#discussion_r57202850>:
>
> > +	"github.com/ethereum/go-ethereum/crypto"
> > +)
> > +
> > +// SignerFn is a signer function callback when a contract requires a method to
> > +// sign the transaction before submission.
> > +type SignerFn func(common.Address, *types.Transaction) (*types.Transaction, error)
> > +
> > +// CallOpts is the collection of options to fine tune a contract call request.
> > +type CallOpts struct {
> > +	Pending bool // Whether to operate on the pending state or the last known one
> > +}
> > +
> > +// TransactOpts is the collection of authorization data required to create a
> > +// valid Ethereum transaction.
> > +type TransactOpts struct {
> > +	Account common.Address // Ethereum account to send the transaction from
>
> I still prefer the use of from ;-)
> Guess its a matter of personal preference.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ethereum/go-ethereum/pull/2357/files/67f1d565173e3a5a69d3f6fb2279ba22a839bc68#r57202850>
>

karalabe replied:
Changed it to `From` :)

bas-vk commented on accounts/abi/bind/template.go:
```diff
@@ -0,0 +1,205 @@
+	//
+	// Solidity: {{.Original.String}}
+	func (_{{$.Type}} *{{$.Type}}Transactor) {{.Normalized.Name}}(opts *bind.TransactOpts {{range .Normalized.Inputs}}, {{.Name}} {{bindtype .Type}} {{end}}) (*types.Transaction, error) {
+		return _{{$.Type}}.contract.Transact(opts, "{{.Normalized.Name}}" {{range .Normalized.Inputs}}, {{.Name}}{{end}})
```
Should `{{.Normalized.Name}}` not be `{{.Original.Name}}` when looking up the correct function?

karalabe replied:
Indeed, very nice catch!

karalabe replied:
Fixed and added a test to call every type of method (that actually does catch the issue if I leave it non-fixed).

bas-vk commented:
Would be nice if you could add a link to the wiki page somewhere in the go doc.
I think it is good source to get started.

karalabe commented:
@bas-vk @obscuren I've pushed a fixed version (changes only in last commit https://github.com/ethereum/go-ethereum/pull/2357/commits/f14d8068562adce3a310d3c541b92a77cf5e3089) and rebased on develop to pull in the core changes.

Although the commit seems quite large, it's small, just re-indented the entire template:

 * Regrouped the binding tests so that more things are checked in a single test, reducing the number of bind-compile-test cycles by 10 tests, or approximately to 10 seconds from 28.
 * Fixed the method naming issue @bas-vk found in the template and wrote a full end-to-end test that deploys a contract, interacts with it and reads its state in a simulator to make sure all calls are working properly internally too.
 * Added the wiki page as a package doc comment to `bind` so that it's easier for people to find it if they look at the package.

The only real feature insert in this commit is modifying `abigen` to support binding Solidity contracts (`--sol`). This entails calling the `compiler` package that wraps `solc` to build the contract fist and only then invoke the `bind.Bind` method on it to generate the resulting file. Since a solidity file can contain multiple contracts in one, I also needed to modify `Bind` a bit to actually accept a slice of abis/bytescodes/typenames and generate bindings for each of them. The big modification in the diff is this, which needed an extra `contracts` range loop in the template, which indented the entire thing,

PTAL

bas-vk commented:
:+1

karalabe commented:
@robotally Slaker :P

karalabe commented:
I've squashed the 16 commits into 3 with some meaningful milestones in between them. If all CIs pass, it's merge ready from my part.
